test/common/singleton/threadsafe_singleton_test.cc

```diff
@@ -0,0 +1,94 @@
+#include <mutex>
+
+#include "common/common/thread.h"
+#include "common/singleton/threadsafe_singleton.h"
+
+#include "test/test_common/threadsafe_singleton_injector.h"
+
+#include "gtest/gtest.h"
+
+namespace Envoy {
+
+class TestSingleton {
+public:
+  virtual ~TestSingleton() {}
+
+  virtual void addOne() {
+    std::unique_lock<std::mutex> lock(lock_);
+    ++value_;
+  }
+
+  virtual int value() {
+    std::unique_lock<std::mutex> lock(lock_);
+    return value_;
+  }
+
+protected:
+  std::mutex lock_;
+  int value_{0};
+};
+
+class EvilMathSingleton : public TestSingleton {
+public:
+  EvilMathSingleton() { value_ = -50; }
+  virtual void addOne() {
+    std::unique_lock<std::mutex> lock(lock_);
+    ++value_;
+    ++value_;
+  }
+};
+
+class AddTen {
+public:
+  AddTen() {
+    thread_.reset(new Thread::Thread([this]() -> void { threadRoutine(); }));
+  }
+  ~AddTen() {
+    thread_->join();
+    thread_.reset();
+  }
+
+private:
+  void threadRoutine() {
+    auto& singleton = ThreadSafeSingleton<TestSingleton>::get();
+    for (int i = 0; i < 10; ++i) {
+      singleton.addOne();
+    }
+  }
+  Thread::ThreadPtr thread_;
+};
+
+TEST(ThreadSafeSingleton, BasicCreationAndMutation) {
+  auto& singleton = ThreadSafeSingleton<TestSingleton>::get();
+  EXPECT_EQ(&singleton, &ThreadSafeSingleton<TestSingleton>::get());
+  EXPECT_EQ(0, singleton.value());
+  singleton.addOne();
+  EXPECT_EQ(1, singleton.value());
+
+  {
+    AddTen ten;
+    AddTen twenty;
+    AddTen thirty;
```

Comment:
Not a huge fan of multithreaded unit tests with arbitrary interleavings. These will be hard to replay when tests flake. Would be a bit more deterministic if we created these on different threads and then ran the routines one per thread. For this particular case, I don't think it matter, it's not going to flake, but as a general principle for future.

Reply:
Can you rephrase the request here?  I think to really test multithreadedness you have to have contention which you can't force without trying to do the updates all at once which means good multi-threaded tests should be non-deterministic by design.

Reply:
The point is that tests have more limited value when they can't be repeated on flakes. I would prefer the ability to control the interleaving precisely of thread contention based on provided seed, but I don't think it's worth it for this simple example.